File: room/room-rxjava3/src/test/java/androidx/room/rxjava3/RxRoomTest.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room.rxjava3;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import androidx.arch.core.executor.ArchTaskExecutor;
import androidx.arch.core.executor.testing.CountingTaskExecutorRule;
import androidx.room.InvalidationTracker;
import androidx.room.RoomDatabase;

import org.hamcrest.CoreMatchers;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.functions.Consumer;
import io.reactivex.rxjava3.observers.TestObserver;
import io.reactivex.rxjava3.subscribers.TestSubscriber;

@RunWith(JUnit4.class)
public class RxRoomTest {
    @Rule
    public CountingTaskExecutorRule mExecutor = new CountingTaskExecutorRule();

    private RoomDatabase mDatabase;
    private InvalidationTracker mInvalidationTracker;
    private List<InvalidationTracker.Observer> mAddedObservers = new ArrayList<>();

    @Before
    public void init() {
        mDatabase = mock(RoomDatabase.class);
        mInvalidationTracker = mock(InvalidationTracker.class);
        when(mDatabase.getInvalidationTracker()).thenReturn(mInvalidationTracker);
        when(mDatabase.getQueryExecutor()).thenReturn(ArchTaskExecutor.getIOThreadExecutor());
        doAnswer(invocation -> {
            mAddedObservers.add((InvalidationTracker.Observer) invocation.getArguments()[0]);
            return null;
        }).when(mInvalidationTracker).addObserver(any(InvalidationTracker.Observer.class));
    }

    @Test
    public void basicAddRemove_Flowable() {
        Flowable<Object> flowable = RxRoom.createFlowable(mDatabase, "a", "b");
        verify(mInvalidationTracker, never()).addObserver(any(InvalidationTracker.Observer.class));
        Disposable disposable = flowable.subscribe();
        verify(mInvalidationTracker).addObserver(any(InvalidationTracker.Observer.class));
        assertThat(mAddedObservers.size(), CoreMatchers.is(1));

        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        disposable.dispose();

        verify(mInvalidationTracker).removeObserver(observer);

        disposable = flowable.subscribe();
        verify(mInvalidationTracker, times(2))
                .addObserver(any(InvalidationTracker.Observer.class));
        assertThat(mAddedObservers.size(), CoreMatchers.is(2));
        assertThat(mAddedObservers.get(1), CoreMatchers.not(CoreMatchers.sameInstance(observer)));
        InvalidationTracker.Observer observer2 = mAddedObservers.get(1);
        disposable.dispose();
        verify(mInvalidationTracker).removeObserver(observer2);
    }

    @Test
    public void basicAddRemove_Observable() {
        Observable<Object> observable = RxRoom.createObservable(mDatabase, "a", "b");
        verify(mInvalidationTracker, never()).addObserver(any(InvalidationTracker.Observer.class));
        Disposable disposable = observable.subscribe();
        verify(mInvalidationTracker).addObserver(any(InvalidationTracker.Observer.class));
        assertThat(mAddedObservers.size(), CoreMatchers.is(1));

        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        disposable.dispose();

        verify(mInvalidationTracker).removeObserver(observer);

        disposable = observable.subscribe();
        verify(mInvalidationTracker, times(2))
                .addObserver(any(InvalidationTracker.Observer.class));
        assertThat(mAddedObservers.size(), CoreMatchers.is(2));
        assertThat(mAddedObservers.get(1), CoreMatchers.not(CoreMatchers.sameInstance(observer)));
        InvalidationTracker.Observer observer2 = mAddedObservers.get(1);
        disposable.dispose();
        verify(mInvalidationTracker).removeObserver(observer2);
    }

    @Test
    public void basicNotify_Flowable() {
        String[] tables = {"a", "b"};
        Set<String> tableSet = new HashSet<>(Arrays.asList(tables));
        Flowable<Object> flowable = RxRoom.createFlowable(mDatabase, tables);
        CountingConsumer consumer = new CountingConsumer();
        Disposable disposable = flowable.subscribe(consumer);
        assertThat(mAddedObservers.size(), CoreMatchers.is(1));
        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        assertThat(consumer.mCount, CoreMatchers.is(1));
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(2));
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(3));
        disposable.dispose();
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(3));
    }

    @Test
    public void basicNotify_Observable() {
        String[] tables = {"a", "b"};
        Set<String> tableSet = new HashSet<>(Arrays.asList(tables));
        Observable<Object> observable = RxRoom.createObservable(mDatabase, tables);
        CountingConsumer consumer = new CountingConsumer();
        Disposable disposable = observable.subscribe(consumer);
        assertThat(mAddedObservers.size(), CoreMatchers.is(1));
        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        assertThat(consumer.mCount, CoreMatchers.is(1));
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(2));
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(3));
        disposable.dispose();
        observer.onInvalidated(tableSet);
        assertThat(consumer.mCount, CoreMatchers.is(3));
    }

    @Test
    public void internalCallable_Flowable() throws Exception {
        final AtomicReference<String> value = new AtomicReference<>(null);
        String[] tables = {"a", "b"};
        Set<String> tableSet = new HashSet<>(Arrays.asList(tables));
        final Flowable<String> flowable = RxRoom.createFlowable(
                mDatabase, false, tables, value::get);
        final CountingConsumer consumer = new CountingConsumer();
        final Disposable disposable = flowable.subscribe(consumer);
        drain();
        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        // no value because it is null
        assertThat(consumer.mCount, CoreMatchers.is(0));
        value.set("bla");
        observer.onInvalidated(tableSet);
        drain();
        // get value
        assertThat(consumer.mCount, CoreMatchers.is(1));
        observer.onInvalidated(tableSet);
        drain();
        // get value
        assertThat(consumer.mCount, CoreMatchers.is(2));
        value.set(null);
        observer.onInvalidated(tableSet);
        drain();
        // no value
        assertThat(consumer.mCount, CoreMatchers.is(2));
        disposable.dispose();
    }

    @Test
    public void internalCallable_Observable() throws Exception {
        final AtomicReference<String> value = new AtomicReference<>(null);
        String[] tables = {"a", "b"};
        Set<String> tableSet = new HashSet<>(Arrays.asList(tables));
        final Observable<String> flowable = RxRoom.createObservable(
                mDatabase, false, tables, value::get);
        final CountingConsumer consumer = new CountingConsumer();
        final Disposable disposable = flowable.subscribe(consumer);
        drain();
        InvalidationTracker.Observer observer = mAddedObservers.get(0);
        // no value because it is null
        assertThat(consumer.mCount, CoreMatchers.is(0));
        value.set("bla");
        observer.onInvalidated(tableSet);
        drain();
        // get value
        assertThat(consumer.mCount, CoreMatchers.is(1));
        observer.onInvalidated(tableSet);
        drain();
        // get value
        assertThat(consumer.mCount, CoreMatchers.is(2));
        value.set(null);
        observer.onInvalidated(tableSet);
        drain();
        // no value
        assertThat(consumer.mCount, CoreMatchers.is(2));
        disposable.dispose();
    }

    @Test
    public void exception_Flowable() throws Exception {
        final Flowable<String> flowable = RxRoom.createFlowable(
                mDatabase,
                false,
                new String[]{"a"},
                () -> {
                    throw new Exception("i want exception");
                });
        TestSubscriber<String> subscriber = new TestSubscriber<>();
        flowable.subscribe(subscriber);
        drain();
        subscriber.assertError(
                throwable -> Objects.equals(throwable.getMessage(), "i want exception"));
    }

    @Test
    public void exception_Observable() throws Exception {
        final Observable<String> flowable = RxRoom.createObservable(
                mDatabase,
                false,
                new String[]{"a"},
                () -> {
                    throw new Exception("i want exception");
                });
        TestObserver<String> observer = new TestObserver<>();
        flowable.subscribe(observer);
        drain();
        observer.assertError(
                throwable -> Objects.equals(throwable.getMessage(), "i want exception"));
    }

    private void drain() throws Exception {
        mExecutor.drainTasks(10, TimeUnit.SECONDS);
    }

    private static class CountingConsumer implements Consumer<Object> {
        int mCount = 0;

        @Override
        public void accept(@NonNull Object o) {
            mCount++;
        }
    }
}

File: room/room-rxjava3/src/test/java/androidx/room/rxjava3/RxRoomTest.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.room.rxjava3

import androidx.arch.core.executor.ArchTaskExecutor
import androidx.arch.core.executor.testing.CountingTaskExecutorRule
import androidx.kruth.assertThat
import androidx.room.InvalidationTracker
import androidx.room.RoomDatabase
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.observers.TestObserver
import io.reactivex.rxjava3.subscribers.TestSubscriber
import java.util.concurrent.Callable
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.invocation.InvocationOnMock
import org.mockito.kotlin.any
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

@RunWith(JUnit4::class)
class RxRoomTest {
    @get:Rule var mExecutor = CountingTaskExecutorRule()
    private lateinit var mDatabase: RoomDatabase
    private lateinit var mInvalidationTracker: InvalidationTracker
    private val mAddedObservers: MutableList<InvalidationTracker.Observer> = ArrayList()

    @Before
    fun init() {
        mDatabase = mock()
        mInvalidationTracker = mock()
        whenever(mDatabase.invalidationTracker).thenReturn(mInvalidationTracker)
        whenever(mDatabase.queryExecutor).thenReturn(ArchTaskExecutor.getIOThreadExecutor())
        doAnswer { invocation: InvocationOnMock ->
                mAddedObservers.add(invocation.arguments[0] as InvalidationTracker.Observer)
                null
            }
            .whenever(mInvalidationTracker)
            .addObserver(any())
    }

    @Test
    fun basicAddRemove_Flowable() {
        val flowable = createFlowable(mDatabase, "a", "b")
        verify(mInvalidationTracker, never()).addObserver(any())
        var disposable = flowable.subscribe()
        verify(mInvalidationTracker).addObserver(any())
        assertThat(mAddedObservers.size).isEqualTo(1)
        val observer = mAddedObservers[0]
        disposable.dispose()
        verify(mInvalidationTracker).removeObserver(observer)
        disposable = flowable.subscribe()
        verify(mInvalidationTracker, times(2)).addObserver(any())
        assertThat(mAddedObservers.size).isEqualTo(2)
        assertThat(mAddedObservers[1]).isNotSameInstanceAs(observer)

        val observer2 = mAddedObservers[1]
        disposable.dispose()
        verify(mInvalidationTracker).removeObserver(observer2)
    }

    @Test
    fun basicAddRemove_Observable() {
        val observable = createObservable(mDatabase, "a", "b")
        verify(mInvalidationTracker, never()).addObserver(any())
        var disposable = observable.subscribe()
        verify(mInvalidationTracker).addObserver(any())
        assertThat(mAddedObservers.size).isEqualTo(1)
        val observer = mAddedObservers[0]
        disposable.dispose()
        verify(mInvalidationTracker).removeObserver(observer)
        disposable = observable.subscribe()
        verify(mInvalidationTracker, times(2)).addObserver(any())
        assertThat(mAddedObservers.size).isEqualTo(2)
        assertThat(mAddedObservers[1]).isNotSameInstanceAs(observer)

        val observer2 = mAddedObservers[1]
        disposable.dispose()
        verify(mInvalidationTracker).removeObserver(observer2)
    }

    @Test
    fun basicNotify_Flowable() {
        val tables = arrayOf("a", "b")
        val tableSet: Set<String> = HashSet(listOf(*tables))
        val flowable = createFlowable(mDatabase, *tables)
        val consumer = CountingConsumer()
        val disposable = flowable.subscribe(consumer)
        assertThat(mAddedObservers.size).isEqualTo(1)
        val observer = mAddedObservers[0]
        assertThat(consumer.mCount).isEqualTo(1)
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(2)
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(3)
        disposable.dispose()
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(3)
    }

    @Test
    fun basicNotify_Observable() {
        val tables = arrayOf("a", "b")
        val tableSet: Set<String> = HashSet(listOf(*tables))
        val observable = createObservable(mDatabase, *tables)
        val consumer = CountingConsumer()
        val disposable = observable.subscribe(consumer)
        assertThat(mAddedObservers.size).isEqualTo(1)
        val observer = mAddedObservers[0]
        assertThat(consumer.mCount).isEqualTo(1)
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(2)
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(3)
        disposable.dispose()
        observer.onInvalidated(tableSet)
        assertThat(consumer.mCount).isEqualTo(3)
    }

    @Test
    @Suppress("DEPRECATION")
    fun internalCallable_Flowable() {
        val value = AtomicReference<Any>(null)
        val tables = arrayOf("a", "b")
        val tableSet: Set<String> = HashSet(listOf(*tables))
        val flowable = createFlowable(mDatabase, false, tables, Callable { value.get() })
        val consumer = CountingConsumer()
        val disposable = flowable.subscribe(consumer)
        drain()
        val observer = mAddedObservers[0]
        // no value because it is null
        assertThat(consumer.mCount).isEqualTo(0)
        value.set("bla")
        observer.onInvalidated(tableSet)
        drain()
        // get value
        assertThat(consumer.mCount).isEqualTo(1)
        observer.onInvalidated(tableSet)
        drain()
        // get value
        assertThat(consumer.mCount).isEqualTo(2)
        value.set(null)
        observer.onInvalidated(tableSet)
        drain()
        // no value
        assertThat(consumer.mCount).isEqualTo(2)
        disposable.dispose()
    }

    @Test
    @Suppress("DEPRECATION")
    fun internalCallable_Observable() {
        val value = AtomicReference<Any>(null)
        val tables = arrayOf("a", "b")
        val tableSet: Set<String> = HashSet(listOf(*tables))
        val flowable = createObservable(mDatabase, false, tables, Callable { value.get() })
        val consumer = CountingConsumer()
        val disposable = flowable.subscribe(consumer)
        drain()
        val observer = mAddedObservers[0]
        // no value because it is null
        assertThat(consumer.mCount).isEqualTo(0)
        value.set("bla")
        observer.onInvalidated(tableSet)
        drain()
        // get value
        assertThat(consumer.mCount).isEqualTo(1)
        observer.onInvalidated(tableSet)
        drain()
        // get value
        assertThat(consumer.mCount).isEqualTo(2)
        value.set(null)
        observer.onInvalidated(tableSet)
        drain()
        // no value
        assertThat(consumer.mCount).isEqualTo(2)
        disposable.dispose()
    }

    @Test
    @Suppress("DEPRECATION")
    fun exception_Flowable() {
        val flowable =
            createFlowable<String>(
                mDatabase,
                false,
                arrayOf("a"),
                Callable { throw Exception("i want exception") }
            )
        val subscriber = TestSubscriber<String>()
        flowable.subscribe(subscriber)
        drain()
        subscriber.assertError { throwable: Throwable -> throwable.message == "i want exception" }
    }

    @Test
    @Suppress("DEPRECATION")
    fun exception_Observable() {
        val flowable =
            createObservable<String>(
                mDatabase,
                false,
                arrayOf("a"),
                Callable { throw Exception("i want exception") }
            )
        val observer = TestObserver<String>()
        flowable.subscribe(observer)
        drain()
        observer.assertError { throwable: Throwable -> throwable.message == "i want exception" }
    }

    @Throws(Exception::class)
    private fun drain() {
        mExecutor.drainTasks(10, TimeUnit.SECONDS)
    }

    private class CountingConsumer : Consumer<Any> {
        var mCount = 0

        override fun accept(o: Any) {
            mCount++
        }
    }
}
